### PR TITLE
Internal analytics for matching engine.

### PR DESCRIPTION
**Pull request by chunter3**
successful_prefect_run

No runtime errors occurred during Prefect test run.

### CODE REVIEW

File: pipeline/types/engine.py [hunk 1]
```diff
@@ -48,3 +48,10 @@ class MatchingOutput:
     release: ReleaseTag
     users: List[User]
     matches: List[Match]
+
+
+@dataclass
+class MatchingInternalAnalytics:  # More metrics will be added in the future
+    n_proposed_matches_per_user: List[
+        Tuple[User, int]
+    ]  # The list will be a tuple containing the user and their number of proposed matches
```
Comment (vingkan) [inline pipeline/types/engine.py:57]: What do you think about a dictionary instead of a list of tuples? What are the pros and cons?

File: pipeline/matching/core/engine.py [hunk 2]
```diff
@@ -19,6 +25,9 @@ class UserNotMatchedException(Exception):
 
 
 class MatchingEngine:
+
+    analytics: MatchingInternalAnalytics  # analytics are saved to this variable w/out affecting the matching engine output
```
Comment (vingkan) [inline pipeline/matching/core/engine.py:29]: Move this into the __init__ method, it should be an instance variable instead of a class variable. More about the differences between those two in this article: https://pynative.com/python-class-variables

Also initialize the object.

File: pipeline/types/engine.py (hunk 1 above)
Comment (vingkan) [inline pipeline/types/engine.py:57]: You can also set a default value for fields in dataclasses.

For a list:

```python
name: Type = field(default_factory=list)
```

For a dictionary:

```python
name: Type = field(default_factory=dict)
```

File: pipeline/matching/core/analytics.py [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+        n_proposed_matches_per_user = []
+        for user in inp.users:
+            match_count = 0
+            for generator in self.matching_engine.generators:
+                gen_matches = generator.generate(inp)
+                for rank_match in self.matching_engine.ranker.rank(
+                    inp, gen_matches
+                ):
+                    if user.uid in rank_match.users:
+                        match_count += 1
```
Comment (vingkan) [inline pipeline/matching/core/analytics.py:23]: instead of doing the actual generation and ranking, this component should only be doing analytics.

So what we want is to take the generator of ranked matches as an input, and then include the analytics in the main engine.

Change the input of this method to take `Iterator[Match]` you can also have the `MatchingInput` as an input, but I don't think we need it for anything. Can you think of anything?

Then we can integrate your analytics class to run after line 41 of the core matching engine code. That way, the matches will already be generated and ranked, you can pass them through your analytics code to collect the counts, and then return the generator of matches untouched and the matching engine code can continue as if nothing happened.

**Review by vingkan — CHANGES_REQUESTED** [review]
Excellent start! The biggest change is that we should make this class something that can be integrated into the matching engine without the class itself calling any generators or rankers, it should only handle counting users, and just pass the ranked matches onto the engine to do the matching.

File: pipeline/matching/core/analytics.py (hunk 3 above)
Comment (vingkan) [reply pipeline/matching/core/analytics.py:23]: To be even more clear, this is how the analytics code should integrate into the core matching engine:

```python
    def run(self, inp: MatchingInput) -> MatchingOutput:
        # Generate matches
        proposed: Iterator[Match] = self.generate_matches(inp)
        # Rank matches
        ranked: Iterator[Match] = self.rank_matches(inp, proposed)
        #################
        # This part is where your code goes:
        analyzer = CollectInternalAnalytics
        collected: Iterator[Match] = analyzer.n_proposed_matches_per_user(ranked)
        # Don't forget to change the input on the line below from ranked to collected
        #################
        # Select priority matches and get remaining users
        results = self.select_priority_matches(inp, collected) # Change the input on this line too
        selected: List[Match] = results[0]
        users: Set[UserId] = results[1]
        # Finalize remaining matches
        finalized: List[Match] = self.finalize_matches(inp, selected, users)
```

The reason we want to do it this way is for separation of concerns:
- Analytics code should only be worried about analytics
- Analytics code should not actually run the generators or rankers, just take their output
- We should be able to make some kinds of changes to the logic of the matching engine and still use the analytics code
- Unit tests for analytics code do not need any generators or rankers, just an iterator of matches, they become much simpler to test
- We can also write an integration test that shows how the analyzer and engine work together, that will be a stretch goal, we can pair on that unless you already know how to write it

File: pipeline/matching/core/analytics.py [hunk 4]
```diff
@@ -0,0 +1,27 @@
+from pipeline.matching.core import MatchingEngine
+from pipeline.types import MatchingInput, MatchingInternalAnalytics
+
+
+class CollectInternalAnalytics:
+    def __init__(self, name, matching_engine):
+        self.name: str = name
+        self.matching_engine: MatchingEngine = matching_engine
```
Comment (vingkan) [inline pipeline/matching/core/analytics.py:8]: You don't need either of these properties, this can be an empty constructor.

```python
class CollectInternalAnalytics:
    def __init__(self):
        pass
```

File: pipeline/matching/core/analytics.py [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+
+    def n_proposed_matches_per_user(
+        self, inp: MatchingInput
+    ) -> MatchingInternalAnalytics:
+
+        n_proposed_matches_per_user = []
+        for user in inp.users:
+            match_count = 0
+            for generator in self.matching_engine.generators:
```
Comment (vingkan) [inline pipeline/matching/core/analytics.py:17]: What is the runtime of this code? Can it be improved?

File: pipeline/matching/core/analytics_test.py [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+    )
+    mock_generator = MagicMock()
+    mock_generator.generate = MagicMock(
+        return_value=[
+            Match(users={"A", "B"}),
+            Match(users={"C", "D"}),
+        ]
+    )
+    mock_ranker = MagicMock()
+    mock_ranker.rank = MagicMock(return_value=[Match(users={"A", "D"})])
```
Comment (vingkan) [inline pipeline/matching/core/analytics_test.py:34]: Good job on learning how to use `MagicMock`!

We shouldn't need all this setup, all we need is one iterator of matches as an input, then these tests will be much shorter and easier to understand.

An iterator can be a list, but if you want to really match the input the engine will give you, we should use a generator. You can turn a list into a generator by calling the `iter()` method on it.

For example, this is what your test bodies will look like, much simpler:

```python
matches = [
    Match(users={"A", "B"}),
    Match(users={"C", "D"}),
]
analyzer = CollectInternalAnalytics()
actual = analyzer.n_proposed_matches_per_user(iter(matches))
...
assert actual == expected
```

File: pipeline/matching/core/analytics.py [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+
+
+class CollectInternalAnalytics:
+    def __init__(self, name, matching_engine):
+        self.name: str = name
+        self.matching_engine: MatchingEngine = matching_engine
+
+    def n_proposed_matches_per_user(
+        self, inp: MatchingInput
+    ) -> MatchingInternalAnalytics:
```
Comment (vingkan) [inline pipeline/matching/core/analytics.py:12]: Does this need to be a class with a function or can it just be a function? What do you think are the pros and cons?

File: pipeline/types/engine.py (hunk 1 above)
Comment (chunter3) [reply pipeline/types/engine.py:57]: Now that you mention it, I think a dictionary would work best in this instance. At the time I was more concerned with how I should integrate the user and the associate metric together; and I thought a tuple would be good. However, a dictionary where the user is the key (since all users should be unique anyway) and the desired metric is the value would be more sensible.

File: pipeline/matching/core/engine.py (hunk 2 above)
Comment (chunter3) [reply pipeline/matching/core/engine.py:29]: This has me confused because I was looking over the issue you opened (#224) and in it mentions this: "Create a **class variable** in MatchingEngine (pipeline/matching/core/engine.py) after the generators/rankers/finalizers for the internal analytics object [. . .]." Unless this was a mistake I did create a class variable; but if this should be an instance variable instance, I'll revise accordingly.

File: pipeline/types/engine.py (hunk 1 above)
Comment (chunter3) [reply pipeline/types/engine.py:57]: Yep. I agree. I already revised the dataclass to be a Dict instead of a List. I'll set the default value to be a dict as well.

File: pipeline/matching/core/analytics.py (hunk 3 above)
Comment (chunter3) [reply pipeline/matching/core/analytics.py:23]: I see. I was under the impression I wasn't meant to alter the existing matching engine code or its associated functions. But if this is the case, then I see where I went wrong and have a better idea of what I should do going forward. I will revise accordingly.

File: pipeline/matching/core/analytics.py (hunk 7 above)
Comment (chunter3) [reply pipeline/matching/core/analytics.py:12]: Well...I was thinking that if we are going to include more metrics in the future, then they would fall under this CollectInternalAnalytics class. While it can be its own function, if we do plan to add more metrics it will help in keeping things more organized. Not to mention using a class is a good way to represent a collection of attributes and methods that may be used repeatedly in other places.

File: pipeline/matching/core/analytics_test.py (hunk 6 above)
Comment (chunter3) [reply pipeline/matching/core/analytics_test.py:34]: Wow...that is very simple. Good to know. Just another instance of me slightly overcomplicating the objective. I will revise my test cases accordingly.

File: pipeline/matching/core/analytics.py (hunk 3 above)
Comment (chunter3) [reply pipeline/matching/core/analytics.py:23]: Also don't we need MatchingInput as an input in order to pull all the users? I could be wrong but that was my assumption. I'll add MatchingInput as an input but will revise accordingly if it turns out to be unnecessary.

File: pipeline/matching/core/engine.py [hunk 8]
```diff
@@ -7,6 +7,7 @@
 from pipeline.matching.core.finalizer import MatchFinalizer
 from pipeline.matching.core.generator import MatchGenerator
 from pipeline.matching.core.ranker import MatchRanker
+from pipeline.matching.evaluation.analytics import CollectInternalAnalytics
```
Comment (vingkan) [inline pipeline/matching/core/engine.py:10]: Delete this line since we won't use analytics inside the core engine.

File: pipeline/matching/core/__init__.py [hunk 9]
```diff
@@ -1,3 +1,4 @@
 from pipeline.matching.core.constants import *
 from pipeline.matching.core.engine import *
 from pipeline.matching.core.generator import *
+from pipeline.matching.evaluation.analytics import *
```
Comment (vingkan) [inline pipeline/matching/core/__init__.py:4]: Delete this line, we should not cross-import packages in the `__init__.py` file.

File: pipeline/matching/core/engine_test.py [hunk 10]
```diff
@@ -1,5 +1,7 @@
 from unittest.mock import MagicMock
 
+import pytest
+
```
Comment (vingkan) [inline pipeline/matching/core/engine_test.py:4]: Delete these lines since they are unused.

File: pipeline/types/engine.py [hunk 11]
```diff
@@ -54,3 +54,10 @@ class MatchingOutput:
     users: List[User]
     matches: List[Match]
     internal_data: MatchingInternalData = MatchingInternalData()
+
+
+@dataclass
+class MatchingInternalAnalytics:  # More metrics will be added in the future
+    n_proposed_matches_per_user: Dict = field(
+        default_factory=dict
+    )  # A dictionary where keys -> users and values -> metrics
```
Comment (vingkan) [inline pipeline/types/engine.py:63]: As a coding style point, I think it is easier to read and understand if the comments are on their own line, above the section they refer to. Like this:

```python
@dataclass
class MatchingInternalAnalytics: 
    # A dictionary where keys -> users and values -> metrics
    n_proposed_matches_per_user: Dict = field(
        default_factory=dict
    ) 
    # More metrics will be added in the future
```

Comment (vingkan) [inline pipeline/types/engine.py:63]: Update this comment, the keys are users and the values will be integers.

File: pipeline/types/engine.py [hunk 12]
```diff
@@ -54,3 +54,10 @@ class MatchingOutput:
     users: List[User]
     matches: List[Match]
     internal_data: MatchingInternalData = MatchingInternalData()
+
+
+@dataclass
+class MatchingInternalAnalytics:  # More metrics will be added in the future
```
Comment (vingkan) [inline pipeline/types/engine.py:60]: Now that it's not internal anymore, we can rename this to `MatchingMetrics`, I think metrics is a better description than analytics.

File: pipeline/matching/evaluation/analytics.py [hunk 13]
```diff
@@ -0,0 +1,24 @@
+from typing import Dict, Iterator, List
+
+from pipeline.types import Match, MatchingInput, MatchingInternalAnalytics, User
+
+
+class CollectInternalAnalytics:
```
Comment (vingkan) [inline pipeline/matching/evaluation/analytics.py:6]: How about we rename this to `MatchingMetricsCollector`? It is sometimes helpful for the class name to be a noun and the methods to be verbs.

You can rename the method to be `count_proposed_matches_per_user()` to match the verb format, or just leave it as-is.

File: pipeline/matching/evaluation/analytics.py [hunk 14]
```diff
@@ -0,0 +1,24 @@
+from typing import Dict, Iterator, List
```
Comment (vingkan) [inline pipeline/matching/evaluation/analytics.py:1]: How about we rename these files to be metrics instead of analytics?

```
pipeline/matching/evaluation/metrics_test.py
pipeline/matching/evaluation/metrics.py
```

File: pipeline/matching/evaluation/analytics.py [hunk 15]
```diff
@@ -0,0 +1,24 @@
...
+        analytics: Dict = {}
+        for match in ranked_matches:
+            for user_id in match.users:
+                if user_id in analytics:
+                    analytics[user_id] += 1
+                    continue
+                analytics[user_id] = 1
+        for user in users:
+            if user.uid not in analytics:
+                analytics[user.uid] = 0
```
Comment (vingkan) [inline pipeline/matching/evaluation/analytics.py:23]: Good job! This now runs in O(M + U) where M is the number of matches and U is the number of users. The old code was O(M x U) which is much more inefficient.

File: pipeline/matching/evaluation/analytics.py [hunk 16]
```diff
@@ -0,0 +1,24 @@
...
+
+class CollectInternalAnalytics:
+    def __init__(self):
+        pass
+
+    def n_proposed_matches_per_user(
+        self, users: List[User], ranked_matches: Iterator[Match]
+    ) -> MatchingInternalAnalytics:
+
+        analytics: Dict = {}
```
Comment (vingkan) [inline pipeline/matching/evaluation/analytics.py:14]: Have you heard of `defaultdict` before? It is part of the built-in `collections` module with Python. Might make this code cleaner because it allows for a default value when the key is not yet in the dictionary.

Up to you if you want to look it up and use that!

If you use a `defaultdict`, cast it to a regular dict when you return it, that way other keys don't get added when callers try to check invalid users.

File: pipeline/matching/evaluation/analytics.py [hunk 17]
```diff
@@ -0,0 +1,24 @@
...
+        for match in ranked_matches:
+            for user_id in match.users:
+                if user_id in analytics:
+                    analytics[user_id] += 1
+                    continue
+                analytics[user_id] = 1
+        for user in users:
+            if user.uid not in analytics:
+                analytics[user.uid] = 0
+        return MatchingInternalAnalytics(n_proposed_matches_per_user=analytics)
```
Comment (vingkan) [inline pipeline/matching/evaluation/analytics.py:24]: We don't need to return the whole analytics object here, just the dict of user IDs to counts.

File: pipeline/matching/evaluation/analytics_test.py [hunk 18]
```diff
@@ -0,0 +1,73 @@
...
+    matches = [Match(users={"A", "B"}), Match(users={"C", "D"})]
+    user_analytics = CollectInternalAnalytics()
+    actual = user_analytics.n_proposed_matches_per_user(users, iter(matches))
+    expected = MatchingInternalAnalytics(
+        n_proposed_matches_per_user={"A": 1, "B": 1, "C": 1, "D": 1, "E": 0}
+    )
+
+    assert actual == expected
+
+    # This test varifies the behavior of the internal analyzer w/ a simple case
```
Comment (vingkan) [inline pipeline/matching/evaluation/analytics_test.py:26]: Typo "verifies", also it is easier for other readers if the comment is at the top of the test:

```python
def test_analytics_logic():
    # This test verifies the behavior of the internal analyzer w/ a simple case
```

File: pipeline/matching/evaluation/analytics_test.py [hunk 19]
```diff
@@ -0,0 +1,73 @@
...
+    matches = [Match(users={})]
+    user_analytics = CollectInternalAnalytics()
+    actual = user_analytics.n_proposed_matches_per_user(users, iter(matches))
+    expected = MatchingInternalAnalytics(
+        n_proposed_matches_per_user={"A": 0, "B": 0, "C": 0, "D": 0, "E": 0}
+    )
+
+    assert actual == expected
+
+    # This test varifies the behavior of the internal analyzer when a ranker returns a match w/ no users
```
Comment (vingkan) [inline pipeline/matching/evaluation/analytics_test.py:73]: Good job on the test cases! I think you captured the import edge cases.

There is an edge case where there is a match for a user who is not in the input, we don't have to worry about that because that would be a problem caused by the generator, not by the metrics. Bad input means bad output :)

File: pipeline/matching/core/engine.py (hunk 2 above)
Comment (vingkan) [reply pipeline/matching/core/engine.py]: Yes, I was not careful with my words on the issue. But now it is moot point because we don't need any instance or class variables!

File: pipeline/types/engine.py (hunk 1 above)
Comment (vingkan) [reply pipeline/types/engine.py]: Awesome, that sounds good!

File: pipeline/matching/core/analytics.py (hunk 3 above)
Comment (vingkan) [reply pipeline/matching/core/analytics.py]: After our pairing call, Chris and I realized that it would be much easier to decouple the analytics from the matching engine code, that way they can function separately. The interface between them is that we now return internal data from the matching engine that the analytics can use. Merged in #242.

File: pipeline/matching/core/analytics.py (hunk 5 above)
Comment (vingkan) [reply pipeline/matching/core/analytics.py]: Chris and I discussed how to improve the runtime! Chris was thinking about the case where some users get no matches and how we want to include a zero count for them, but that can be done more efficiently as well. Good job improving the performance of this, Chris!

File: pipeline/matching/core/analytics_test.py (hunk 6 above)
Comment (vingkan) [reply pipeline/matching/core/analytics_test.py]: Excellent, great job refactoring the tests!

File: pipeline/matching/core/analytics.py (hunk 7 above)
Comment (vingkan) [reply pipeline/matching/core/analytics.py]: I agree with you, let's keep it as a class then!

One thing to think about is merge conflicts, sometimes if multiple people are writing new metrics in the same class, they might create conflicts for each other that slow down changes and reviews. But, we can learn how to handle those.

**Review by vingkan — CHANGES_REQUESTED** [review]
Excellent work Chris! I left all the comments needed for clean-up, once those are done, I will approve and you can merge.

**Review by vingkan — APPROVED** [review]
Excellent work, Chris! This is a very clean pull request with efficient code and good tests.
